File: app/src/main/java/com/example/ftpserver/MainActivity.java
package com.example.ftpserver;

import android.os.Bundle;
import android.widget.Button;
import android.widget.TextView;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Environment;
import android.widget.Toast;

import org.apache.ftpserver.FtpServer;
import org.apache.ftpserver.FtpServerFactory;
import org.apache.ftpserver.ftplet.Authority;
import org.apache.ftpserver.ftplet.UserManager;
import org.apache.ftpserver.listener.ListenerFactory;
import org.apache.ftpserver.usermanager.PropertiesUserManagerFactory;
import org.apache.ftpserver.usermanager.impl.BaseUser;
import org.apache.ftpserver.usermanager.impl.WritePermission;
import org.apache.ftpserver.usermanager.impl.ConcurrentLoginPermission;
import org.apache.ftpserver.usermanager.impl.TransferRatePermission;
import org.apache.ftpserver.ConnectionConfigFactory;

import java.net.InetAddress;
import java.net.NetworkInterface;
import java.net.Inet4Address;
import java.util.ArrayList;
import java.util.Enumeration;
import java.util.List;
import java.io.File;

public class MainActivity extends AppCompatActivity {
    private FtpServer server;
    private TextView statusText;
    private Button toggleButton;
    private static final int PORT = 2121;
    private static final int PERMISSION_REQUEST_CODE = 1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        statusText = findViewById(R.id.statusText);
        toggleButton = findViewById(R.id.toggleButton);
        
        toggleButton.setOnClickListener(v -> {
            if (server != null && !server.isStopped()) {
                stopFtpServer();
            } else {
                checkPermissionAndStartServer();
            }
        });
    }

    private void checkPermissionAndStartServer() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)
                != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this,
                    new String[]{
                            Manifest.permission.WRITE_EXTERNAL_STORAGE,
                            Manifest.permission.READ_EXTERNAL_STORAGE
                    },
                    PERMISSION_REQUEST_CODE);
        } else {
            startFtpServer();
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == PERMISSION_REQUEST_CODE) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                startFtpServer();
            } else {
                Toast.makeText(this, "需要存储权限才能运行FTP服务器", Toast.LENGTH_LONG).show();
            }
        }
    }

    private void startFtpServer() {
        try {
            FtpServerFactory serverFactory = new FtpServerFactory();
            ListenerFactory factory = new ListenerFactory();
            
            factory.setPort(PORT);
            
            PropertiesUserManagerFactory userManagerFactory = new PropertiesUserManagerFactory();
            UserManager userManager = userManagerFactory.createUserManager();
            
            BaseUser anonymousUser = new BaseUser();
            anonymousUser.setName("anonymous");
            anonymousUser.setPassword("");
            
            // 直接使用外部存储根目录（储存卡）
            File homeDir = Environment.getExternalStorageDirectory();
            if (homeDir != null && homeDir.exists()) {
                anonymousUser.setHomeDirectory(homeDir.getAbsolutePath());
            } else {
                // 如果外部存储不可用，使用应用私有目录作为备选
                homeDir = getExternalFilesDir(null);
                if (homeDir != null) {
                    anonymousUser.setHomeDirectory(homeDir.getAbsolutePath());
                } else {
                    anonymousUser.setHomeDirectory(getFilesDir().getAbsolutePath());
                }
            }
            
            List<Authority> authorities = new ArrayList<>();
            authorities.add(new WritePermission());
            authorities.add(new ConcurrentLoginPermission(10, 10));
            authorities.add(new TransferRatePermission(0, 0));
            anonymousUser.setAuthorities(authorities);
            
            userManager.save(anonymousUser);
            serverFactory.setUserManager(userManager);
                
            ConnectionConfigFactory configFactory = new ConnectionConfigFactory();
            configFactory.setAnonymousLoginEnabled(true);
            serverFactory.setConnectionConfig(configFactory.createConnectionConfig());
            
            serverFactory.addListener("default", factory.createListener());
            
            server = serverFactory.createServer();
            server.start();
            
            String ipAddress = getLocalIpAddress();
            statusText.setText("FTP服务器运行中\n" +
                    "IP: " + ipAddress + "\n" +
                    "端口: " + PORT + "\n" +
                    "根目录: " + homeDir.getAbsolutePath() + "\n" +
                    "支持匿名访问");
            toggleButton.setText("停止服务器");
            
        } catch (Exception e) {
            e.printStackTrace();
            statusText.setText("启动服务器失败: " + e.getMessage());
        }
    }

    private void stopFtpServer() {
        if (server != null) {
            server.stop();
            statusText.setText("FTP服务器已停止");
            toggleButton.setText("启动服务器");
        }
    }

    private String getLocalIpAddress() {
        try {
            for (Enumeration<NetworkInterface> en = NetworkInterface.getNetworkInterfaces(); 
                 en.hasMoreElements();) {
                NetworkInterface intf = en.nextElement();
                for (Enumeration<InetAddress> enumIpAddr = intf.getInetAddresses(); 
                     enumIpAddr.hasMoreElements();) {
                    InetAddress inetAddress = enumIpAddr.nextElement();
                    if (!inetAddress.isLoopbackAddress() && inetAddress instanceof Inet4Address) {
                        return inetAddress.getHostAddress();
                    }
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return "未知";
    }
} 

File: kt/app/src/main/java/com/example/ftpserver/MainActivity.kt
package com.example.ftpserver

import android.Manifest
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.os.PowerManager
import android.provider.Settings
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.localbroadcastmanager.content.LocalBroadcastManager

class MainActivity : AppCompatActivity() {
    private lateinit var statusText: TextView
    private lateinit var toggleButton: Button
    private lateinit var btnClient: Button

    companion object {
        private const val PERMISSION_REQUEST_CODE = 1
        private const val MANAGE_STORAGE_REQUEST_CODE = 2
        private const val NOTIFICATION_PERMISSION_REQUEST_CODE = 3
    }

    /** 接收 FtpService 的状态变化，刷新界面 */
    private val stateReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            val running = intent?.getBooleanExtra(FtpService.EXTRA_RUNNING, false) ?: false
            val message = intent?.getStringExtra(FtpService.EXTRA_MESSAGE) ?: ""
            updateUi(running, message)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        statusText = findViewById(R.id.statusText)
        toggleButton = findViewById(R.id.toggleButton)
        btnClient = findViewById(R.id.btnClient)

        toggleButton.setOnClickListener {
            if (FtpService.isRunning) {
                FtpService.stop(this)
            } else {
                checkPermissionAndStartServer()
            }
        }

        btnClient.setOnClickListener {
            val intent = Intent(this, FtpClientActivity::class.java)
            startActivity(intent)
        }

        LocalBroadcastManager.getInstance(this).registerReceiver(
            stateReceiver,
            IntentFilter(FtpService.ACTION_STATE_CHANGED)
        )

        // 应用启动时自动请求权限
        checkAndRequestPermissions()
        requestNotificationPermission()
    }

    override fun onResume() {
        super.onResume()
        updateUi(FtpService.isRunning, FtpService.statusMessage)
    }

    override fun onDestroy() {
        LocalBroadcastManager.getInstance(this).unregisterReceiver(stateReceiver)
        super.onDestroy()
    }

    private fun updateUi(running: Boolean, message: String) {
        statusText.text = message
        toggleButton.text = if (running) "停止服务器" else "启动服务器"
    }

    // Android 13+ 前台服务通知需要通知权限
    private fun requestNotificationPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            if (ContextCompat.checkSelfPermission(
                    this,
                    Manifest.permission.POST_NOTIFICATIONS
                ) != PackageManager.PERMISSION_GRANTED
            ) {
                ActivityCompat.requestPermissions(
                    this,
                    arrayOf(Manifest.permission.POST_NOTIFICATIONS),
                    NOTIFICATION_PERMISSION_REQUEST_CODE
                )
            }
        }
    }

    // 请求忽略电池优化，避免锁屏后被系统冻结
    private fun requestIgnoreBatteryOptimizations() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            val powerManager = getSystemService(Context.POWER_SERVICE) as PowerManager
            if (!powerManager.isIgnoringBatteryOptimizations(packageName)) {
                try {
                    val intent = Intent(
                        Settings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS,
                        Uri.parse("package:$packageName")
                    )
                    startActivity(intent)
                    Toast.makeText(this, "请允许后台运行，锁屏后FTP才不会被系统关闭", Toast.LENGTH_LONG).show()
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }
    }

    // 应用启动时检查并请求权限
    private fun checkAndRequestPermissions() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            // Android 11+ 需要请求所有文件访问权限
            if (!Environment.isExternalStorageManager()) {
                try {
                    val intent = Intent(Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION).apply {
                        data = Uri.parse("package:$packageName")
                    }
                    startActivityForResult(intent, MANAGE_STORAGE_REQUEST_CODE)
                    Toast.makeText(this, "请授予所有文件访问权限以使用FTP服务器", Toast.LENGTH_LONG).show()
                } catch (e: Exception) {
                    val intent = Intent(Settings.ACTION_MANAGE_ALL_FILES_ACCESS_PERMISSION)
                    startActivityForResult(intent, MANAGE_STORAGE_REQUEST_CODE)
                    Toast.makeText(this, "请授予所有文件访问权限以使用FTP服务器", Toast.LENGTH_LONG).show()
                }
            }
        } else {
            // Android 10 及以下使用传统权限请求
            if (ContextCompat.checkSelfPermission(
                    this,
                    Manifest.permission.WRITE_EXTERNAL_STORAGE
                ) != PackageManager.PERMISSION_GRANTED
            ) {
                ActivityCompat.requestPermissions(
                    this,
                    arrayOf(
                        Manifest.permission.WRITE_EXTERNAL_STORAGE,
                        Manifest.permission.READ_EXTERNAL_STORAGE
                    ),
                    PERMISSION_REQUEST_CODE
                )
            }
        }
    }

    private fun checkPermissionAndStartServer() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            // Android 11+ 检查所有文件访问权限
            if (Environment.isExternalStorageManager()) {
                startFtpServer()
            } else {
                checkAndRequestPermissions()
            }
        } else {
            // Android 10 及以下使用传统权限
            if (ContextCompat.checkSelfPermission(
                    this,
                    Manifest.permission.WRITE_EXTERNAL_STORAGE
                ) != PackageManager.PERMISSION_GRANTED
            ) {
                ActivityCompat.requestPermissions(
                    this,
                    arrayOf(
                        Manifest.permission.WRITE_EXTERNAL_STORAGE,
                        Manifest.permission.READ_EXTERNAL_STORAGE
                    ),
                    PERMISSION_REQUEST_CODE
                )
            } else {
                startFtpServer()
            }
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == PERMISSION_REQUEST_CODE) {
            if (grantResults.isNotEmpty()) {
                val allPermissionsGranted = grantResults.all { it == PackageManager.PERMISSION_GRANTED }
                if (allPermissionsGranted) {
                    startFtpServer()
                } else {
                    // 处理权限被拒绝的情况
                    Toast.makeText(this, "需要所有权限才能启动服务器", Toast.LENGTH_SHORT).show()
                }
            }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == MANAGE_STORAGE_REQUEST_CODE) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                if (Environment.isExternalStorageManager()) {
                    Toast.makeText(this, "已获得所有文件访问权限", Toast.LENGTH_SHORT).show()
                } else {
                    Toast.makeText(this, "需要所有文件访问权限才能使用FTP服务器", Toast.LENGTH_LONG).show()
                }
            }
        }
    }

    private fun startFtpServer() {
        // 启动前台服务，由服务持有 WakeLock/WifiLock，锁屏后继续运行
        requestIgnoreBatteryOptimizations()
        FtpService.start(this)
    }
}
